### PR TITLE
Remove option to paste rich text from Android EditText context menu

### PR DESCRIPTION
**Pull request by fabriziobertoglio1987**
## Summary:

Text is copy pasted as rich text on Android TextInput instead of Plain Text.

### What is the root cause of that problem?

Android EditText and iOS UITextField/UITextView have different copy/paste behavior.
- Android TextInput copies/pastes rich text
- iOS UITextField copies/pastes plain text.

| iOS (react-native)   | Android (react-native) |
| ----------- | ----------- |
| <video src="https://user-images.githubusercontent.com/24992535/249170968-8fde35f0-a53c-4c5c-bd89-ee822c08eadf.mp4" width="350" />      | <video src="https://user-images.githubusercontent.com/24992535/249171968-bf0915a0-4060-4586-b267-7c2b463d76f6.mov" width="350" />       |

### What changes do you think we should make in order to solve the problem?

The issue is a bug in react-native https://github.com/facebook/react-native/issues/31442:

1) The JavaScript TextInput and ReactEditText Android state are not in sync
2) The TextInput Android Native state over-rides the JavaScript state.
3) onChangeText passes a plain text string to JavaScript, not rich text (text with spans and styles).

More info at https://github.com/Expensify/App/issues/21411#issuecomment-1611591137

The solution consists of:

1) **Over-riding onTextContextMenuItem in ReactEditText to copy/paste plain text instead of rich-text** (https://stackoverflow.com/a/45319485/7295772).
2) **Removing the `Paste as plaintext` option from the insert and selection context menu** 

fixes https://github.com/facebook/react-native/issues/31442

## Changelog:

[ANDROID] [FIXED] - Remove option to paste rich text from Android EditText context menu

#### Reproducing the issue on Android 

https://user-images.githubusercontent.com/24992535/249185416-76f8a687-1aca-4dc9-9abe-3d73d6e2893c.mp4

#### Fixing the issue on Android
Sourcecode https://github.com/fabriziobertoglio1987/text-input-cursor-flickering-android/blob/fix-copy-paste/app/src/main/java/com/example/myapplication/CustomEditText.java

https://user-images.githubusercontent.com/24992535/249486339-95449bb9-71b6-430c-8207-f5672f034fa9.mp4

#### Testing the solution on React Native

https://github.com/Expensify/App/assets/24992535/b302237b-99e5-44a2-996d-8bc50bbbc95c

### CODE REVIEW

**Comment by fabriziobertoglio1987** [comment]
Building the main branch with head https://github.com/facebook/react-native/commit/06668fcbacd750771f1d53cce829dc55e86f3f3c triggers a runtime when opening the RNTester Text examples:

https://github.com/facebook/react-native/assets/24992535/a8971f2b-f19a-4acc-a22e-fb7c1e34df86

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java [hunk 1]
```diff
@@ -269,6 +271,37 @@ public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
...
+        id = android.R.id.pasteAsPlainText;
+      } else {
+        ClipboardManager clipboard =
+            (ClipboardManager) getContext().getSystemService(Context.CLIPBOARD_SERVICE);
+        ClipData clip = clipboard.getPrimaryClip();
+        if (clip != null) {
+          for (int i = 0; i < clip.getItemCount(); i++) {
+            final CharSequence paste;
+            final CharSequence text = clip.getItemAt(i).coerceToText(getContext());
+            paste = (text instanceof Spanned) ? text.toString() : text;
```
Comment (fabriziobertoglio1987) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java:291]: refactor this

**Comment by NickGerleman** [comment]
Could you explain why you think the current behavior is a bug? At a glance, overriding a context menu command to paste rich text, to instead coerce to plain text, seems at odds with platform expectations.

**Comment by fabriziobertoglio1987** [comment]
Thanks @NickGerleman 

>Could you explain why you think the current behavior is a bug?

**FIRST REASON**: It was reported as a react-native TextInput Issue in different chat applications:
- Zulip Mobile https://github.com/zulip/zulip-mobile/issues/4660#issuecomment-1203296441
>It's already the case that the only data TextInput actually exposes to the app is the plain text -- so the fact that it lets the user paste formatted text, and shows that formatting back, is basically a bug in RN because there's no way the app can end up respecting that formatting.

https://github.com/zulip/zulip-mobile/issues/4660#issuecomment-951338651
>As described on [this CZO thread](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/problem.20with.20bulleted.20list/near/1270772), this issue can lead to a very awkward user experience.
- Expensify https://github.com/Expensify/App/issues/21411
- React Native Issue: https://github.com/facebook/react-native/issues/31442

**SECOND REASON**: React Native Controlled TextInput Component does not support paste with rich text.

- Native and JavaScript state are not in sync. 
- The rich text pasted in the EditText over-rides the style of a JavaScript Controlled TextInput.

**Example of a use case**: The user pastes rich text in a JavaScript Controlled TextInput.

**Expected behaviour**: 
- The JavaScript Controlled TextInput manages the TextInput state and over-rides the Native Android State. The JavaScript TextInput and Text style props set the native text style.
- onChangeText API is used to update the TextInput internal state and keep it sync with the updates from native.

**Actual behaviour**: 
- The JavaScript TextInput and Text style props do not change the style of the text. 
- onChangeText passes to JavaScript callback the plain text instead of rich text. It is not possible to keep the two states in sync.

https://user-images.githubusercontent.com/24992535/249502554-0f3a48c6-38ee-4847-be17-2aabebf52377.mp4

more info in comment https://github.com/Expensify/App/issues/21411#issuecomment-1611591137

> At a glance, overriding a context menu command to paste rich text, to instead coerce to plain text, seems at odds with platform expectations.

The fix is published as two PRs:

1) The first PR [`over-rides onTextContextMenuItem in ReactEditText to copy/paste plain text instead of rich-text` #38189](https://github.com/facebook/react-native/pull/38189)
2) The second PR [`removes the option paste as plain text` #38210](https://github.com/facebook/react-native/pull/38210)

The final result is that the option `paste as plain text` is removed, while `paste` becomes `paste as plain text`.

| Before/After iOS   | Before Android |
| ----------- | ----------- |
| <video src="https://user-images.githubusercontent.com/24992535/249170968-8fde35f0-a53c-4c5c-bd89-ee822c08eadf.mp4" width="350" />      | <video src="https://user-images.githubusercontent.com/24992535/249171968-bf0915a0-4060-4586-b267-7c2b463d76f6.mov" width="350" />       |

| After Android    |
| ----------- |
| <video src="https://github.com/Expensify/App/assets/24992535/b302237b-99e5-44a2-996d-8bc50bbbc95c" width="1000" />      |

**Review by NickGerleman — APPROVED** [review]
Could we combine https://github.com/facebook/react-native/pull/38210 into this so that we get both changes at once? Otherwise this makes sense to me, if the application can never have visibility into the rich text.

**Comment by fabriziobertoglio1987** [comment]
https://github.com/facebook/react-native/pull/38189#pullrequestreview-1542055466 

Thanks a lot @NickGerleman. I combined [#38210](https://github.com/facebook/react-native/pull/38210) with [#38189](https://github.com/facebook/react-native/pull/38189). I tested the PR again and did not detect any issues.

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java [hunk 2]
```diff
@@ -269,6 +303,36 @@ public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
...
+      } else {
+        ClipboardManager clipboard =
+            (ClipboardManager) getContext().getSystemService(Context.CLIPBOARD_SERVICE);
+        ClipData clip = clipboard.getPrimaryClip();
+        if (clip != null) {
+          for (int i = 0; i < clip.getItemCount(); i++) {
+            final CharSequence text = clip.getItemAt(i).coerceToText(getContext());
+            final CharSequence paste = (text instanceof Spanned) ? text.toString() : text;
+            if (paste != null) {
+              ClipData clipData = ClipData.newPlainText("rebase_copy", text);
```
Comment (NickGerleman) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java:324]: nit: why this label?

Comment (fabriziobertoglio1987) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java]: Thanks @NickGerleman I changed the label to null as implemented in TextView#onTextContextMenuItem for ID_COPY.

https://github.com/aosp-mirror/platform_frameworks_base/blob/27c5bed4fa582fa934da55528c12af414167de8d/core/java/android/widget/TextView.java#L12928

Relevant https://github.com/aosp-mirror/platform_frameworks_base/blob/27c5bed4fa582fa934da55528c12af414167de8d/core/java/android/content/ClipData.java#L849

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java [hunk 3]
```diff
@@ -269,6 +303,36 @@ public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
...
+              ClipboardManager manager =
+                  (ClipboardManager) getContext().getSystemService(Context.CLIPBOARD_SERVICE);
+              manager.setPrimaryClip(clipData);
```
[the remark this answers is not in the document]
Comment (fabriziobertoglio1987) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java]: Thanks @NickGerleman. I changed the implementation to restore the previous clip after calling super.onTextContextMenuItem(id).

| Before    | After |
| ----------- | ----------- |
| <image src="https://user-images.githubusercontent.com/24992535/256245526-caf8a45f-ed57-4bb3-8110-83454dbbe0c5.gif" width="350" />      | <image src="https://user-images.githubusercontent.com/24992535/256246789-8cec95e9-76d5-453e-941d-2883dac00459.gif" width="350" />       |

**Pull request closed without merging**